File: typescript/docs/examples/browser-use/browser/browser-fingerprint-persistence.ts
/**
 * Integration test for browser fingerprint persistence functionality.
 * This test verifies that browser fingerprint can be persisted
 * across sessions using the same ContextId and FingerprintContextId.
 */

// @ts-nocheck
import { AgentBay, CreateSessionParams } from 'wuying-agentbay-sdk';
import { BrowserOption, BrowserFingerprint, BrowserContext, BrowserFingerprintContext } from 'wuying-agentbay-sdk';
import { chromium } from 'playwright';

// Global variables for persistent context and fingerprint context
let persistentContext: any = null;
let persistentFingerprintContext: any = null;

function getTestApiKey(): string {
  const apiKey = process.env.AGENTBAY_API_KEY;
  if (!apiKey) {
    throw new Error("AGENTBAY_API_KEY environment variable not set");
  }
  return apiKey;
}

function isWindowsUserAgent(userAgent: string): boolean {
  if (!userAgent) {
    return false;
  }
  const userAgentLower = userAgent.toLowerCase();
  const windowsIndicators = [
    'windows nt',
    'win32',
    'win64',
    'windows',
    'wow64'
  ];
  return windowsIndicators.some(indicator => userAgentLower.includes(indicator));
}

async function runAsFirstTime(): Promise<void> {
  console.log("====================");
  console.log("Run as first time");
  console.log("====================");

  const apiKey = getTestApiKey();
  const agentBay = new AgentBay({ apiKey });

  // Create a browser context for first time
  const sessionContextName = `test-browser-context-${Date.now()}`;
  const contextResult = await agentBay.context.get(sessionContextName, true);
  if (!contextResult.success || !contextResult.context) {
    console.log("Failed to create browser context");
    return;
  }

  persistentContext = contextResult.context;
  console.log(`Created browser context: ${persistentContext.name} (ID: ${persistentContext.id})`);

  // Create a browser fingerprint context for first time
  const fingerprintContextName = `test-browser-fingerprint-${Date.now()}`;
  const fingerprintContextResult = await agentBay.context.get(fingerprintContextName, true);
  if (!fingerprintContextResult.success || !fingerprintContextResult.context) {
    console.log("Failed to create fingerprint context");
    return;
  }

  persistentFingerprintContext = fingerprintContextResult.context;
  console.log(`Created fingerprint context: ${persistentFingerprintContext.name} (ID: ${persistentFingerprintContext.id})`);

  // Create session with BrowserContext and FingerprintContext
  console.log(`Creating session with browser context ID: ${persistentContext.id} ` +
              `and fingerprint context ID: ${persistentFingerprintContext.id}`);
  
  const fingerprintContext = new BrowserFingerprintContext(persistentFingerprintContext.id);
  const browserContext = new BrowserContext(persistentContext.id, true, undefined, fingerprintContext);

  const params = new CreateSessionParams()
    .withImageId("browser_latest")
    .withBrowserContext(browserContext);

  const sessionResult = await agentBay.create(params);
  if (!sessionResult.success || !sessionResult.session) {
    console.log(`Failed to create first session: ${sessionResult.errorMessage}`);
    return;
  }

  const session = sessionResult.session;
  console.log(`First session created with ID: ${session.sessionId}`);

  // Get browser object and generate fingerprint for persistence
  async function firstSessionOperations(): Promise<void> {
    // Initialize browser with fingerprint persistent enabled and set fingerprint generation options
    const browserOption: BrowserOption = {
      useStealth: true,
      fingerprintPersistent: true,
      fingerprint: {
        devices: ["desktop"],
        operatingSystems: ["windows"],
        locales: ["zh-CN"]
      } as BrowserFingerprint
    };

    const initSuccess = await session.browser.initializeAsync(browserOption);
    if (!initSuccess) {
      console.log("Failed to initialize browser");
      return;
    }
    console.log("First session browser initialized successfully");

    // Get endpoint URL
    const endpointUrl = await session.browser.getEndpointUrl();
    if (!endpointUrl) {
      console.log("Failed to get browser endpoint URL");
      return;
    }
    console.log(`First session browser endpoint URL: ${endpointUrl}`);

    // Connect with playwright, test first session fingerprint
    console.log("Opening https://httpbin.org/user-agent and test user agent...");
    const browser = await chromium.connectOverCDP(endpointUrl);
    const context = browser.contexts().length > 0 ? browser.contexts()[0] : await browser.newContext();

    const page = await context.newPage();
    await page.goto("https://httpbin.org/user-agent", { timeout: 60000 });
    const response = await page.evaluate(() => JSON.parse(document.body.textContent));
    const userAgent = response["user-agent"];
    console.log("user_agent =", userAgent);
    
    const isWindows = isWindowsUserAgent(userAgent);
    if (!isWindows) {
      console.log("Failed to get windows user agent");
      return;
    }

    await context.close();
    console.log("First session browser fingerprint check completed");
  }

  // Run first session operations
  await firstSessionOperations();

  // Delete first session with syncContext=true
  console.log("Deleting first session with syncContext=true...");
  const deleteResult = await agentBay.delete(session, { syncContext: true });
  console.log(`First session deleted successfully (RequestID: ${deleteResult.requestId})`);
}

async function runAsSecondTime(): Promise<void> {
  console.log("====================");
  console.log("Run as second time");
  console.log("====================");

  const apiKey = getTestApiKey();
  const agentBay = new AgentBay({ apiKey });

  // Create second session with same browser context and fingerprint context
  console.log(`Creating second session with same browser context ID: ${persistentContext.id} ` +
              `and fingerprint context ID: ${persistentFingerprintContext.id}`);
  
  const fingerprintContext = new BrowserFingerprintContext(persistentFingerprintContext.id);
  const browserContext = new BrowserContext(persistentContext.id, true, undefined, fingerprintContext);

  const params = new CreateSessionParams()
    .withImageId("browser_latest")
    .withBrowserContext(browserContext);

  const sessionResult = await agentBay.create(params);
  if (!sessionResult.success || !sessionResult.session) {
    console.log(`Failed to create second session: ${sessionResult.errorMessage}`);
    return;
  }

  const session = sessionResult.session;
  console.log(`Second session created with ID: ${session.sessionId}`);

  // Get browser object and check if second session fingerprint is the same as first session
  async function secondSessionOperations(): Promise<void> {
    // Initialize browser with fingerprint persistent enabled but not specific fingerprint generation options
    const browserOption: BrowserOption = {
      useStealth: true,
      fingerprintPersistent: true
    };

    const initSuccess = await session.browser.initializeAsync(browserOption);
    if (!initSuccess) {
      console.log("Failed to initialize browser in second session");
      return;
    }
    console.log("Second session browser initialized successfully");

    // Get endpoint URL
    const endpointUrl = await session.browser.getEndpointUrl();
    if (!endpointUrl) {
      console.log("Failed to get browser endpoint URL in second session");
      return;
    }
    console.log(`Second session browser endpoint URL: ${endpointUrl}`);

    // Connect with playwright and test second session fingerprint
    const browser = await chromium.connectOverCDP(endpointUrl);
    const context = browser.contexts().length > 0 ? browser.contexts()[0] : await browser.newContext();
    const page = await context.newPage();
    await page.goto("https://httpbin.org/user-agent", { timeout: 60000 });
    const response = await page.evaluate(() => JSON.parse(document.body.textContent));
    const userAgent = response["user-agent"];
    console.log("user_agent =", userAgent);
    
    const isWindows = isWindowsUserAgent(userAgent);
    if (!isWindows) {
      console.log("Failed to get windows user agent in second session");
      return;
    }
    console.log("SUCCESS: fingerprint persisted correctly!");

    await context.close();
    console.log("Second session browser fingerprint check completed");
  }

  // Run second session operations
  await secondSessionOperations();

  // Delete second session with syncContext=true
  console.log("Deleting second session with syncContext=true...");
  const deleteResult = await agentBay.delete(session, { syncContext: true });
  console.log(`Second session deleted successfully (RequestID: ${deleteResult.requestId})`);
}

async function main(): Promise<void> {
  /**
   * Test browser fingerprint persist across sessions with the same browser and fingerprint context.
   */
  try {
    await runAsFirstTime();
    await new Promise(resolve => setTimeout(resolve, 3000)); // Wait 3 seconds
    await runAsSecondTime();
  } catch (error) {
    console.error("Error in main function:", error);
  }
}

if (require.main === module) {
  main().catch(console.error);
}

File: python/docs/examples/browser-use/browser/browser_fingerprint_persistence.py
#!/usr/bin/env python3
"""
Integration test for browser fingerprint persistence functionality.
This test verifies that browser fingerprint can be persisted
across sessions using the same ContextId and FingerprintContextId.
"""

import asyncio
import os
import time
import unittest
from agentbay import AgentBay
from agentbay.session_params import CreateSessionParams, BrowserContext
from agentbay.browser.browser import BrowserOption, BrowserFingerprint, BrowserFingerprintContext
from playwright.async_api import async_playwright

# Global variables for persistent context and fingerprint context
persistent_context = None
persistent_fingerprint_context = None

def get_test_api_key():
    """Get API key for testing"""
    api_key = os.environ.get("AGENTBAY_API_KEY")
    if not api_key:
        raise unittest.SkipTest("AGENTBAY_API_KEY environment variable not set")
    return api_key


def is_windows_user_agent(user_agent: str) -> bool:
    if not user_agent:
        return False
    user_agent_lower = user_agent.lower()
    windows_indicators = [
        'windows nt',
        'win32',
        'win64',
        'windows',
        'wow64'
    ]
    return any(indicator in user_agent_lower for indicator in windows_indicators)


def run_as_first_time():
    """Run as first time"""
    print("="*20)
    print("Run as first time")
    print("="*20)
    global persistent_context, persistent_fingerprint_context
    api_key = os.environ.get("AGENTBAY_API_KEY")
    if not api_key:
        print("Error: AGENTBAY_API_KEY environment variable not set")
        return

    agent_bay = AgentBay(api_key)

    # Create a browser context for first time
    session_context_name = f"test-browser-context-{int(time.time())}"
    context_result = agent_bay.context.get(session_context_name, True)
    if not context_result.success or not context_result.context:
        print("Failed to create browser context")
        return

    persistent_context = context_result.context
    print(f"Created browser context: {persistent_context.name} (ID: {persistent_context.id})")

    # Create a browser fingerprint context for first time
    fingerprint_context_name = f"test-browser-fingerprint-{int(time.time())}"
    fingerprint_context_result = agent_bay.context.get(fingerprint_context_name, True)
    if not fingerprint_context_result.success or not fingerprint_context_result.context:
        print("Failed to create fingerprint context")
        return
    
    persistent_fingerprint_context = fingerprint_context_result.context
    print(f"Created fingerprint context: {persistent_fingerprint_context.name} (ID: {persistent_fingerprint_context.id})")


    # Create session with BrowserContext and FingerprintContext
    print(f"Creating session with browser context ID: {persistent_context.id} "
            f"and fingerprint context ID: {persistent_fingerprint_context.id}")
    fingerprint_context = BrowserFingerprintContext(persistent_fingerprint_context.id)
    browser_context = BrowserContext(persistent_context.id, auto_upload=True, fingerprint_context=fingerprint_context)
    params = CreateSessionParams(
        image_id="browser_latest",
        browser_context=browser_context
    )

    session_result = agent_bay.create(params)
    if not session_result.success or not session_result.session:
        print(f"Failed to create first session: {session_result.error_message}")
        return

    session = session_result.session
    print(f"First session created with ID: {session.session_id}")

    # Get browser object and generate fingerprint for persistence
    async def first_session_operations():
        # Initialize browser with fingerprint persistent enabled and set fingerprint generation options
        browser_option = BrowserOption(
            use_stealth=True,
            fingerprint_persistent=True,
            fingerprint=BrowserFingerprint(
                devices=["desktop"],
                operating_systems=["windows"],
                locales=["zh-CN"],
            ),
        )
        init_success = await session.browser.initialize_async(browser_option)
        if not init_success:
            print("Failed to initialize browser")
            return
        print("First session browser initialized successfully")

        # Get endpoint URL
        endpoint_url = session.browser.get_endpoint_url()
        if not endpoint_url:
            print("Failed to get browser endpoint URL")
            return
        print(f"First session browser endpoint URL: {endpoint_url}")

        # Connect with playwright, test first session fingerprint
        print("Opening https://httpbin.org/user-agent and test user agent...")
        async with async_playwright() as p:
            browser = await p.chromium.connect_over_cdp(endpoint_url)
            context = browser.contexts[0] if browser.contexts else await browser.new_context()

            page = await context.new_page()
            await page.goto("https://httpbin.org/user-agent", timeout=60000)
            response = await page.evaluate("() => JSON.parse(document.body.textContent)")
            user_agent = response["user-agent"]
            print("user_agent =", user_agent)
            is_windows = is_windows_user_agent(user_agent)
            if not is_windows:
                print("Failed to get windows user agent")
                return

            await context.close()
            print("First session browser fingerprint check completed")

    # Run first session operations
    asyncio.run(first_session_operations())

    # Delete first session with syncContext=True
    print("Deleting first session with syncContext=True...")
    delete_result = agent_bay.delete(session, sync_context=True)
    print(f"First session deleted successfully (RequestID: {delete_result.request_id})")


def run_as_second_time():
    """Run as second time"""
    print("="*20)
    print("Run as second time")
    print("="*20)
    global persistent_context, persistent_fingerprint_context
    api_key = os.environ.get("AGENTBAY_API_KEY")
    if not api_key:
        print("Error: AGENTBAY_API_KEY environment variable not set")
        return

    agent_bay = AgentBay(api_key)

    # Create second session with same browser context and fingerprint context
    print(f"Creating second session with same browser context ID: {persistent_context.id} "
            f"and fingerprint context ID: {persistent_fingerprint_context.id}")
    fingerprint_context = BrowserFingerprintContext(persistent_fingerprint_context.id)
    browser_context = BrowserContext(persistent_context.id, auto_upload=True, fingerprint_context=fingerprint_context)
    params = CreateSessionParams(
        image_id="browser_latest",
        browser_context=browser_context
    )
    session_result = agent_bay.create(params)
    if not session_result.success or not session_result.session:
        print(f"Failed to create second session: {session_result.error_message}")
        return

    session = session_result.session
    assert session is not None  # Type narrowing for linter
    print(f"Second session created with ID: {session.session_id}")

    # Get browser object and check if second session fingerprint is the same as first session
    async def second_session_operations():
        # Initialize browser with fingerprint persistent enabled but not specific fingerprint generation options
        browser_option = BrowserOption(
            use_stealth=True,
            fingerprint_persistent=True,
        )
        init_success = await session.browser.initialize_async(browser_option)
        if not init_success:
            print("Failed to initialize browser in second session")
            return
        print("Second session browser initialized successfully")

        # Get endpoint URL
        endpoint_url = session.browser.get_endpoint_url()
        if not endpoint_url:
            print("Failed to get browser endpoint URL in second session")
            return
        print(f"Second session browser endpoint URL: {endpoint_url}")

        # Connect with playwright and test second session fingerprint
        async with async_playwright() as p:
            browser = await p.chromium.connect_over_cdp(endpoint_url)
            context = browser.contexts[0] if browser.contexts else await browser.new_context()
            page = await context.new_page()
            await page.goto("https://httpbin.org/user-agent", timeout=60000)
            response = await page.evaluate("() => JSON.parse(document.body.textContent)")
            user_agent = response["user-agent"]
            print("user_agent =", user_agent)
            is_windows = is_windows_user_agent(user_agent)
            if not is_windows:
                print("Failed to get windows user agent in second session")
                return
            print(f"SUCCESS: fingerprint persisted correctly!")

            await context.close()
            print("Second session browser fingerprint check completed")

    # Run second session operations
    asyncio.run(second_session_operations())

    # Delete second session with syncContext=True
    print("Deleting second session with syncContext=True...")
    delete_result = agent_bay.delete(session, sync_context=True)
    print(f"Second session deleted successfully (RequestID: {delete_result.request_id})")


def main():
    """Test browser fingerprint persist across sessions with the same browser and fingerprint context."""
    run_as_first_time()
    time.sleep(3)
    run_as_second_time()

if __name__ == "__main__":
    main()
